### DISCUSSION TITLE
Configuring Default Browser for Shiny

### DISCUSSION THREAD

**Discussion by keatonwilson** [posit-dev/positron#4568] [Q&A]
Hi there, 

Loving the new IDE and decided to take a stab at a Shiny for Python project (normally an R user here) to learn try and learn two things simultaneously! I'm hoping to get some help configuring the default browser that an app opens in. 

**Step 1:** Run `shiny run` in the terminal
**Step 2:** Helpful dialogue box pops up at the top asking how I want to open the app (see screenshot)
<img width="642" alt="Screenshot 2024-09-03 at 4 33 55 PM" src="https://github.com/user-attachments/assets/31a74d3b-5a21-4976-b103-94d0e9370b34">
**Step 3:** I click on 'Configure default opener... which opens up by settings.json
**Step 4:** settings.json opens to a new property, but I can't figure out how to make it open with Chrome (I'm on a Mac). Abbreviated json below: 

```
{
    "workbench.colorTheme": "Default Dark Modern",
    "python.defaultInterpreterPath": "a/path/that/is/blocked/out/for/anonymity",
    "git.autofetch": true,
    "workbench.externalUriOpeners": {
    
     ## Config points me to here!

    },
    "editor.cursorSmoothCaretAnimation": "on",
    "editor.fontLigatures": true,
    "editor.fontFamily": "Fira Code",
    "editor.rulers": [
        80, 100, 120
    ],
    "editor.defaultFormatter": "esbenp.prettier-vscode",
    "[python]": {
        "editor.defaultFormatter": "ms-python.black-formatter"
    },
}
```

Has anyone else had success with getting this set up? Thanks in advance!

**Comment by juliasilge** [comment]
In terms of editing the JSON, sometimes you might have better luck using the UI version of editing your settings:

![Screenshot 2024-09-04 at 10 02 55 AM](https://github.com/user-attachments/assets/d5c67b5b-266e-4ab0-ba6c-92e17db58d21)

The JSON that creates is like this:

```json
    "workbench.externalUriOpeners": {
        "http": "positron.viewer"
    }
```

However, I don't think doing this will avoid you being prompted to choose among the options, since VS Code and Positron both already provide default openers. [Take a look here](https://code.visualstudio.com/updates/v1_53#_external-uri-opener):

>        // This opener has default priority for this URI.
>        // This will result in the user being prompted since VS Code always has
>        // its own default opener.

I think the best way to think about this is that you already have several default openers (your external browser, Positron's Viewer, VS Code's Simple Browser). You could add _another_ default opener, but then you would just have more to choose from; it doesn't become the default choice over the other defaults.

Let me know if you have further questions!

**Comment by francisbarton** [comment]
I have also been struggling with this. I think there's a conflict between what I expect(ed) the setting to do, and what it is actually doing. I am just adding this comment here in case it is helpful at all to the OP or anyone else finding this.

So, based on Julia's response, what I am now understanding is that the `"workbench.externalUriOpeners"` setting is a way for the user to add in an _extra_ browser option to the list that pops up. (But maybe I am wrong about this).

However my original understanding of what would happen if I set this, was that the "Configure default opener..." menu option would allow me to specify a particular browser or option that I _always_ want to open http links with, thereby skipping the "How would you like to open..." dropdown step.

> you already have several default openers (your external browser, Positron's Viewer, VS Code's Simple Browser). You could add another default opener, but then you would just have more to choose from; it doesn't become the default choice over the other defaults.

This explanation from Julia is helpful in terms of understanding what is happening here, and why even when I set the `"workbench.externalUriOpeners"` setting I still get the dropdown. (This was unexpected and perplexing to me).

But my comment/suggestion would be that the wording here is potentially unhelpful. As a user, if I set a default option, it makes sense to me to assume that Positron will then use that default option without asking me every time. The phrase "several default openers" doesn't make sense to me, as I would understand a 'default' to be a single option out of several, you can't have multiple defaults as that is not what the word means.

Ultimately, I want to be able to tell Positron to always open my site in my external browser, not to ask me every time. I never want to open it in the Viewer pane or the simple browser. I am not sure how to achieve this.

**Reply by juliasilge** [reply]
> what I am now understanding is that the `"workbench.externalUriOpeners"` setting is a way for the user to add in an extra browser option to the list that pops up. (But maybe I am wrong about this).

You are correct, and I certainly agree that "default" is far from an ideal wording here!

I think the feature you may be more interested in is https://github.com/posit-dev/positron/issues/8450. There is a workaround described there that will work for most Shiny apps (although not for apps you run from the Positron console).
